xl_auth/public/forms.py

```diff
@@ -61,7 +61,16 @@ def validate(self):
 
         user = User.get_by_email(self.username.data)
         if user:
-            return True
+            active_resets = PasswordReset.get_active_resets_for_email(user.email)
```

Comment:
I think this is a step away from the previous approach where we have *users* and *collections* at the center. Each of them relates to a bunch of other stuffs, like permissions and clients and tokens and other users and roles and password resets. When not resolved directly from the relation, they have utility methods to filter among related entities. E.g.

```python
class User(UserMixin, SurrogatePK, Model):
    """A user of the app."""
    # ...

    permissions = relationship('Permission', back_populates='user',
                               foreign_keys='Permission.user_id', lazy='joined')

    # ...

    def has_any_permission_for(self, collection):
        """Check for any permission on a specific collection."""
        return any([permission for permission in self.permissions
                    if permission.collection == collection])

    # ...

    # ...

    def get_cataloging_admin_permissions(self):
        """Return all cataloging admin permissions for this user."""
        return [perm for perm in self.permissions if perm.cataloging_admin]
```

I think a simpler and better maintainable approach would be adding this logic alongside *get_subset_of_permission* and friends. E.g.

```python
class PasswordReset(SurrogatePK, Model):
    """Password reset token for a user."""

   # ...

    created_at = Column(db.DateTime, default=datetime.utcnow, nullable=False)

    # ...

    @hybrid_property
    def is_recent(self):
        """Check if reset was recently created."""
        return self.created_at > (datetime.utcnow() - timedelta(hours=2))

# ...

class User(UserMixin, SurrogatePK, Model):
    """A user of the app."""

    # ...

    password_resets = relationship('PasswordReset', back_populates='user')
    
    # ...

    def get_active_and_recent_password_resets(self):
        """Return active and recent password resets for the user."""
        return list(filter(lambda _: _.is_active and _.is_recent, self.password_resets))

```

This way, form logic could be more like

```python
        # ...
        if user:
            if len(user.get_active_and_recent_password_resets()) >= 2:
                self.username.errors.append(_('Chill down and wait')
                return False
            else:
                return True
```

Also means neater and better unittests for the models, methinks.

What do you think?